src/shmem_node_util.c

```diff
@@ -0,0 +1,92 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "shmem_internal.h"
+#include "shmem_node_util.h"
+
+
+char *shmem_internal_location_array = NULL;
+
+static int node_util_is_initialized = 0;
+
+
+int shmem_node_util_init(void) {
+    int ret;
+
+    if (!node_util_is_initialized) {
+        shmem_internal_location_array = malloc(sizeof(char) * shmem_internal_num_pes);
+        if (NULL == shmem_internal_location_array) {
+            RAISE_ERROR_STR("Error: out of memory when allocating node_util location array");
+            return 1;
+        }
+
+        memset(shmem_internal_location_array, -1, shmem_internal_num_pes);
+
+        ret = shmem_runtime_put("nodename", shmem_internal_nodename(), strlen(shmem_internal_nodename())+1);
+        if (ret !=0) {
+            RAISE_ERROR_MSG("shmem_node_util_init failed during nodename store to KVS: (%d)", ret);
+            return ret;
+        }
+
+        node_util_is_initialized = 1;
+    } else {
+        RAISE_WARN_STR("Attemping to initialize shmem_node_util more than once");
+    }
+
+    return 0;
+}
+
+
+inline
+void shmem_node_util_set_node_rank(int pe, int node_rank) {
+
+    shmem_internal_location_array[pe] = node_rank;
+    return;
+}
+
+inline
+int shmem_node_util_get_rank_same_node(int pe) {
+
+#ifdef USE_ON_NODE_COMMS
+    return shmem_internal_location_array[pe];
+#elif defined(USE_MEMCPY)
+    return pe == shmem_internal_my_pe ? 0 : -1;
+#else
+    return -1;
+#endif
```

Comment:
Copy/paste from `GET_RANK_SAME_NODE` == 😿

Reply:
Haha, sad kitty indeed.  I hesitated to use this function instead of the macro, because it's called in all the `shmem_internal` RMAs and AMOs in `shmem_comm.h`, on the critical path.  Do you think it's fine to call this inlined version throughout?  I'd guess it would be fine, but I'm glad it's come up in code review cause I'm not 100% sure.

Reply:
Sure, just move it to the header so it's inlined.